lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectDatumReader.java

```diff
@@ -121,6 +123,18 @@ protected Object readArray(Object old, Schema expected, ResolvingDecoder in)
       @SuppressWarnings("unchecked")
       Collection<Object> c = (Collection<Object>) array;
       return readCollection(c, expectedType, l, in);
+    } else if (array instanceof Map 
+               && ReflectData.isMapWithNonStringKeysSchema(expected)) {
```

Comment:
I think the Schema condition here should be validated in the block or removed. If the object is a `Map`, then the schema must match. In the case where the schema doesn't match that's an error that should throw an `AvroRuntimeException`. Right now, if the object doesn't match the expected Schema, the code assumes that the object is a java array, which it may know is false.

If the object is a `Map`, then it was created (or validated) by `newArray(Object, int, Schema)`. That can be overridden, so I think a check and throw is the right solution.